debian/samsredir.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH sams2redir 1 "January 25, 2008"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
sams2redir \- SAMS simple redirector for squid
.SH SYNOPSIS
.B sams2redir
.RI [ options ] " files" ...
.br
.B bar
.RI [ options ] " files" ...
.SH DESCRIPTION
This manual page documents briefly the
.B sams2redir
command.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBsams2redir\fP is a program that...
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.SH SEE ALSO
.BR samsparser (1),
.BR samsdaemon (1).
.br
.SH AUTHOR
sams2redir was written by SAMS Team.
.PP
This manual page was written by Pavel Vinogradov <Pavel.Vinogradov@nixdev.net>,
for the Debian project (but may be used by others).


NEWS
История версий БД
   База данных имеет свою версию, которая хранится в websettings.s_version
   а) БД считается "актуальной" если первые 3 числа ее версии сопадают с теми-же числами версии пакета.
      Это означает что приложения работают в полнофункциональном режиме.
   б) БД считается "допустимой" если первое и второе число сопадают с теми-же числами версии пакета.
      Это означает что приложения могут работать, но некоторые функции будут недоступны.
   в) БД считается "тестовой", если второе и третье число равны 9.
      Может быть использовано в целях разработки, тестов и т.д.

   1.9.9 - первая тестовая версия
   2.9.9 - добавлено применение модулей sysinfo
           CREATE TABLE `sysinfo` (
             `s_proxy_id` INT              NOT NULL ,
             `s_name`     VARCHAR( 50 )    NOT NULL ,
             `s_version`  VARCHAR( 10 )    NOT NULL ,
             `s_author`   VARCHAR( 30 )        NULL DEFAULT 'anonymous',
             `s_info`     VARCHAR( 1024 )  NOT NULL DEFAULT 'not available',
             `s_date`     DATETIME         NOT NULL ,
             `s_status`   INT              NOT NULL
           ) ENGINE = MYISAM;
           UPDATE websettings SET s_version='2.9.9';
   3.9.9 - добавлена функция перенаправления
           ALTER TABLE `redirect` ADD `s_dest` VARCHAR( 128 ) NULL ;
   4.9.9 - добавлена функция перевода пользователя в другой шаблон при превышении лимита
           ALTER TABLE `shablon` ADD `s_shablon_id2` BIGINT( 20 ) UNSIGNED NULL ;

081116
  Вторичный шаблон пользователя, баннерорезка, добавление комментариев,
  оптимизация работы с группами url адресов
  багфиксы: 
  1) в движке MySQL если числовое значение в БД равно NULL, вылетали демоны
  2) при перечитывании настроек список плагинов системной информации не очищался
081125
  Отправка письма администратору при блокировке пользователя (или переводе во временный шаблон)
081128
  Принимаются во внимание 2 параметра прокси сервера: регистр букв в домене и имени пользователя.
081129
  1) Перевод описания таблиц из формата doc в формат html с частичной проверкой актуальности.
  2) Документация по исходникам собирается и устанавливается, если найден doxygen.
     Скорей всего rpm после этого компилироваться не будет.
081130
  Начало написания документации по установке, настройке и использованию.
081203
  Исправление ошибки выполнения make install на FreeBSD (find: unknown option -iwholename)
  Небольшое дополнение документации
  Документация по исходникам теперь не собирается по умолчанию, для сборки необходимо указать
    опцию --with-srcdocs во время configure
081205
  Учитывается proxy.s_usedomain при создании acl, поиске и создании пользователей.
  Исправлена ошибка при создании БД. Не учитывалось что изменилась таблица shablon.
  Дополнение документации.
081207
  При автоматическом создании пользователя, он не вносился в squid.conf до следующего реконфигурирования.
  Не создавалась таблица sysinfo.
  Обновлена документация.
  Исправлены мелкие ошибки при создании пакета для RedHat.
081209
  Удаление линка wbinfo из репозитария, т.к. более не используется и вносит смуту при построении пакетов.
  Параметры proxy.s_bigd, proxy.s_bigu и proxy.s_squidbase теперь корректно отображаются в web и обновляются в БД.
081217
  Добавлена информация, выводимая на экран, в конце этапа configure.
  Корректная работа команды make uninstall
  Обновление документации (раздел требований)
  Fixed #78
081220
  Исправлена и расширена настройка авторизации OpenLDAP и загрузка пользователей. Требует доработки.
090328
  Доделан механизм регистрации/разрегистрации и автоматического удаления классов типа DBquery.
090329
  Счетчики пользователей после окончания периода ограничения очищаются только если в настройках прокси
  установлен флаг "Автоматически очищать счетчики трафика пользователей"
090407
  В samsdaemon и samsredir учитывается уровень отладочных сообщений, установленный в БД, но только если он
  не был явно задан из командной строки. При нестандартном периоде ограничения трафика в шаблоне, обновляется
  дата следующей очистки, но пока только в оперативной памяти (в БД не изменяется). Незначительное обновление
  документации.
090418
  Иправлена ошибка в определении IP адреса по доменному имени.
  Исправлена несвоевременная разрегистрация MYSQLQuery.
  Добавлено использование внутреннего кеша DNS ответов
090419
  Дата следующей очистки счетчиков у шаблонов с нестандартным периодом теперь обновляется в БД.
  Исправлена регистрация/разрегистрация PgQuery.
  Обновлена документация исходных кодов.
090425
  Обновлена документация исходных кодов. Добавлено описание отличий версии 1 и 2.
090428
  Оптимизирован поиск пользователей. Отключен кеш DNS ответов (необходимо использование TTL ответов)
090429
  Немного подправлено построение rpm пакетов.
090502
  Исправлено отображение статуса пользователя и ограничение трафика в веб интерфейсе.
  При удалении временного интервала в шаблоне теперь не переходит на редактирование другого шаблона.
  Правильно устанавливается значение "Запрет доступа ко всем url" при редактировании шаблона.
090503
  Период лимита шаблона 1 день.
  Правильно устанавливается значение уровня отладочных сообщений при редактировании прокси.
  Добавлено управление системными плагинами. Чтобы заработало, необходимо модифицировать БД.
  TRUNCATE TABLE sysinfo;
  ALTER TABLE  sysinfo ADD s_row_id SERIAL PRIMARY KEY FIRST;
090504
  Добавлена поддержка ограничения по расширениям файлов (пока только для редиректора)
  Теперь в разделы acl и http_access в squid.conf ничего не пишется, если используется встроенный редиректор.
090506
  Исправлена LDAP авторизация пользователя в веб интерфейсе
090507
  Исправлен импорт пользователей из LDAP
  Исправлено редактирование списков регулярных выражений.
  Теперь принимается и частичное совпадение url с регулярным выражением
090511
  Значительно переделан интерфейс редактирования настроек прокси сервера.

trunk/src/samsredir.cpp
/***************************************************************************
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include <getopt.h>
#include <stdlib.h>

#include "config.h"
#include "configmake.h"

#include "dbconn.h"
#include "samsconfig.h"
#include "debug.h"
#include "samsuserlist.h"
#include "samsuser.h"
#include "proxy.h"
#include "localnetworks.h"
//#include "grouplist.h"
#include "templatelist.h"
#include "template.h"
#include "urlgrouplist.h"
#include "tools.h"

/**
 *  Выводит список опций командной строки с кратким описанием
 */
void usage ()
{
  cout << endl;
  cout << "NAME" << endl;
  cout << "    sams2redir - redirector." << endl;
  cout << endl;
  cout << "SYNOPSIS" << endl;
  cout << "    sams2redir [COMMAND] [OPTION]..." << endl;
  cout << endl;
  cout << "COMMANDS" << endl;
  cout << "    -h, --help" << endl;
  cout << "                Show this help screen and exit." << endl;
  cout << "    -V, --version" << endl;
  cout << "                Print program version number and exit." << endl;
  cout << endl;
  cout << "Mandatory arguments to long options are mandatory for short options too." << endl;
  cout << "The following options are available:" << endl;
  cout << endl;
  cout << "OPTIONS" << endl;
  cout << "    -v, --verbose" << endl;
  cout << "                Produce more output." << endl;
  cout << "    -d, --debug=LEVEL" << endl;
  cout << "                Produce lots of debugging information depend on LEVEL." << endl;
  cout << "                For more information about possible LEVEL values refer to a developer." << endl;
  cout << "    -l, --logger=LOGGER" << endl;
  cout << "                Set engine for messages output. Possible values for LOGGER are: console, syslog, file." << endl;
  cout << "                In case of file you can set filename for output (DEFAULT: samsparser.log)." << endl;
  cout << "                E.g. -l syslog" << endl;
  cout << "                E.g. -l file:/path/to/file" << endl;
  cout << "    -C, --config=FILE" << endl;
  cout << "                Use config file FILE." << endl;
  cout << endl;
}

/**
 *  Выводит версию программы и немного информации
 */
void version ()
{
  cout << "sams2redir " << VERSION << endl;
  cout << "Written by anonymous." << endl;
  cout << endl;
  cout << "This program comes with NO WARRANTY; not even for MERCHANTABILITY" << endl;
  cout << "or FITNESS FOR A PARTICULAR PURPOSE." << endl;
}

int main (int argc, char *argv[])
{
  int parse_errors = 0;
  int c;
  int err;
  uint dbglevel_cmd = 0;
  string optname = "";
  bool verbose = false;
  string log_engine = "";
  string config_file = SYSCONFDIR;
  config_file += "/sams2.conf";

  static struct option long_options[] = {
    {"help",     0, 0, 'h'},     // Показывает справку по опциям командной строки и завершает работу
    {"version",  0, 0, 'V'},     // Показывает версию программы и завершает работу
    {"verbose",  0, 0, 'v'},     // Устанавливает режим многословности
    {"debug",    1, 0, 'd'},     // Устанавливает уровень отладки
    {"logger",   1, 0, 'l'},     // Устанавливает движок вывода сообщений
    {"config",   1, 0, 'C'},     // Использовать альтернативный конфигурационный файл
    {0, 0, 0, 0}
  };

  while (1)
    {
      int option_index = 0;

      c = getopt_long (argc, argv, "hVvd:t:l:C:", long_options, &option_index);
      if (c == -1)              // no more options
        break;
      switch (c)
        {
        case 0:
          optname = long_options[option_index].name;
          break;
        case 'h':
          usage ();
          exit (0);
          break;
        case 'V':
          version ();
          exit (0);
          break;
        case 'v':
          verbose = true;
          break;
        case 'd':
          if (sscanf (optarg, "%d", &dbglevel_cmd) != 1)
            dbglevel_cmd = 0;
          break;
        case 'l':
          log_engine = optarg;
          break;
        case 'C':
          config_file = optarg;
          break;
        case '?':
          break;
        default:
          printf ("?? getopt return character code 0%o ??\n", c);
        }
    }

  Logger::setSender("sams2redir");
  SamsConfig::useFile (config_file);

  // Сначала прочитаем конфигурацию, параметры командной строки
  // имеют приоритет, потому анализируются позже
  SamsConfig::reload ();

  uint dbglevel;
  int dbglevel_db = SamsConfig::getInt (defDEBUG, err);
  if ((dbglevel_cmd == 0) && (err == ERR_OK) )
    {
      dbglevel = (uint)dbglevel_db;
    }
  else
    {
      dbglevel = dbglevel_cmd;
    }

  Logger::setEngine (log_engine);
  Logger::setVerbose (verbose);
  Logger::setDebugLevel (dbglevel);

  if (parse_errors > 0)
    {
      usage ();
      exit (parse_errors);
    }


  DBConn *conn = NULL;

  conn = SamsConfig::newConnection ();
  if (!conn)
    {
      ERROR ("Unable to create connection.");
      return 1;
    }

  if (!conn->connect ())
    {
      delete conn;
      return 1;
    }

  SAMSUserList::useConnection (conn);
  LocalNetworks::useConnection (conn);
  //GroupList::useConnection (conn); // not used anywhere
  Proxy::useConnection (conn);
  TemplateList::useConnection (conn);
  UrlGroupList::useConnection (conn);

  SAMSUserList::reload ();
  LocalNetworks::reload ();
  //GroupList::reload (); // not used anywhere
  Proxy::reload ();
  TemplateList::reload ();
  UrlGroupList::reload();

  // select t.s_shablon_id, r.s_type, u.s_url
  // from shablon t, sconfig t_r, redirect r, url u
  // where t_r.s_shablon_id=t.s_shablon_id
  //       and t_r.s_redirect_id=r.s_redirect_id
  //       and u.s_redirect_id=r.s_redirect_id;
  //delete conn;
  //conn = NULL;

//  basic_stringstream < char >mess;

//  mess << "Started with pid " << pid << ".";

//  Logger::addLog(Logger::LK_DAEMON, mess.str());


  char line[2048];
  vector < string > fields;
  vector < string > source;
  string destination;
  SAMSUser *usr = NULL;
  Template *tpl = NULL;
  while (true)
    {
      cin.getline(line, sizeof(line));

      // входная строка имеет вид
      // URL ip-address/fqdn ident method
      Split (line, "\n\t ", fields);

      if (fields.size () == 0)
        break;

      INFO ("Input: " << line);

      // Мы незнаем что такое попалось, но на всякий случай ничего менять не будем
      if (fields.size () < 4)
        {
          INFO ("Reason: Invalid fields count: " << fields.size());
          INFO ("Output: " << line);
          cout << line << endl << flush;
          continue;
        }

      Split (fields[1], "/", source);

      // url считается локальным и неважно какой пользователь обратился, разрешаем доступ
      if (LocalNetworks::isLocalUrl(fields[0]))
        {
          INFO ("Reason: Url is local");
          INFO ("Output: " << line);
	  cout << line << endl << flush;
          continue;
        }

      usr = Proxy::findUser (source[0], url_decode(fields[2]));

      // Пользователь не найден, блокируем доступ
      if (!usr)
        {
          INFO ("Reason: User not found");
          if (fields[2] != "-")
            {
              INFO ("Output: " << Proxy::getDenyAddr () << "/blocked.php?action=usernotfound&id=" << fields[2] << " " << fields[1] << " " << fields[2] << " " << fields[3]);
              cout << Proxy::getDenyAddr () << "/blocked.php?action=usernotfound&id=" << fields[2];
              cout << " " << fields[1] << " " << fields[2] << " " << fields[3] << endl << flush;
            }
          else
            {
              INFO ("Output: " << Proxy::getDenyAddr () << "/blocked.php?action=usernotfound&id=" << source[0] << " " << fields[1] << " " << fields[2] << " " << fields[3]);
              cout << Proxy::getDenyAddr () << "/blocked.php?action=usernotfound&id=" << source[0];
              cout << " " << fields[1] << " " << fields[2] << " " << fields[3] << endl << flush;
            }
          continue;
        }

      DEBUG(DEBUG_REDIR, "[" << __FUNCTION__ << "] Found user: " << *usr);

      if ( (usr->getEnabled () != SAMSUser::STAT_ACTIVE) && (usr->getEnabled () != SAMSUser::STAT_LIMITED) )
        {
          INFO ("Reason: User not active (disabled or blocked)");
          INFO ("Output: " << Proxy::getDenyAddr () << "/blocked.php?action=userdisabled&id=" << *usr << " " << fields[1] << " " << fields[2] << " " << fields[3]);
          cout << Proxy::getDenyAddr () << "/blocked.php?action=userdisabled&id=" << *usr;
          cout << " " << fields[1] << " " << fields[2] << " " << fields[3] << endl << flush;
          continue;
        }

      // нарушена целостность БД (отсутствует шаблон пользователя), блокируем доступ
      tpl = TemplateList::getTemplate (usr->getCurrentTemplateId ());
      if (!tpl)
        {
          INFO ("Reason: User's template not found");
          INFO ("Output: " << Proxy::getDenyAddr () << "/blocked.php?action=templatenotfound&id=" << *usr << " " << fields[1] << " " << fields[2] << " " << fields[3]);
          cout << Proxy::getDenyAddr () << "/blocked.php?action=templatenotfound&id=" << *usr;
          cout << " " << fields[1] << " " << fields[2] << " " << fields[3] << endl << flush;
          continue;
        }

      // Если url существует в белом списке, разрешаем доступ
      if ( tpl->isUrlWhitelisted (fields[0]) )
        {
          INFO ("Reason: In white list");
          INFO ("Output: " << line);
	  cout << line << endl << flush;
          continue;
        }

      // Если url существует в черном списке, блокируем доступ
      if ( tpl->isUrlBlacklisted (fields[0]) )
        {
          INFO ("Reason: In black list");
          INFO ("Output: " << Proxy::getDenyAddr () << "/blocked.php?action=urldenied&id=" << *usr << " " << fields[1] << " " << fields[2] << " " << fields[3]);
          cout << Proxy::getDenyAddr () << "/blocked.php?action=urldenied&id=" << *usr;
          cout << " " << fields[1] << " " << fields[2] << " " << fields[3] << endl << flush;
          continue;
        }

      // Если url существует в списке регулярных выражений, блокируем доступ
      if ( tpl->isUrlMatchRegex (fields[0]) )
        {
          INFO ("Reason: In regular expression list");
          INFO ("Output: " << Proxy::getDenyAddr () << "/blocked.php?action=urldenied&id=" << *usr << " " << fields[1] << " " << fields[2] << " " << fields[3]);
          cout << Proxy::getDenyAddr () << "/blocked.php?action=urldenied&id=" << *usr;
          cout << " " << fields[1] << " " << fields[2] << " " << fields[3] << endl << flush;
          continue;
        }

      // Если в url присутствует запрещенное расширение файла, блокируем доступ
      if ( tpl->isUrlHasFileExt (fields[0]) )
        {
          INFO ("Reason: Has disallowed file extension");
          INFO ("Output: " << Proxy::getDenyAddr () << "/blocked.php?action=urldenied&id=" << *usr << " " << fields[1] << " " << fields[2] << " " << fields[3]);
          cout << Proxy::getDenyAddr () << "/blocked.php?action=urldenied&id=" << *usr;
          cout << " " << fields[1] << " " << fields[2] << " " << fields[3] << endl << flush;
          continue;
        }

      // Если url в текущее время не разрешен, блокируем доступ
      if ( tpl->isTimeDenied (fields[0]) )
        {
          INFO ("Reason: Denied due to time restrictions");
          INFO ("Output: " << Proxy::getDenyAddr () << "/blocked.php?action=timedenied&id=" << *usr << " " << fields[1] << " " << fields[2] << " " << fields[3]);
          cout << Proxy::getDenyAddr () << "/blocked.php?action=timedenied&id=" << *usr;
          cout << " " << fields[1] << " " << fields[2] << " " << fields[3] << endl << flush;
          continue;
        }

      if ( tpl->getAllDeny () )
        {
          INFO ("Reason: Denied to all and not whitelisted");
          INFO ("Output: " << Proxy::getDenyAddr () << "/blocked.php?action=urldenied&id=" << *usr << " " << fields[1] << " " << fields[2] << " " << fields[3]);
          cout << Proxy::getDenyAddr () << "/blocked.php?action=urldenied&id=" << *usr;
          cout << " " << fields[1] << " " << fields[2] << " " << fields[3] << endl << flush;
          continue;
        }

      destination = tpl->modifyUrl (fields[0]);
      if (!destination.empty ())
        {
          INFO ("Reason: Redirected to another location");
          INFO ("Output: " << destination << " " << fields[1] << " " << fields[2] << " " << fields[3]);
          cout << destination;
          cout << " " << fields[1] << " " << fields[2] << " " << fields[3] << endl << flush;
          continue;
        }

      // Все проверки пройдены успешно, разрешаем доступ
      INFO ("Reason: Access granted");
      INFO ("Output: " << line);
      cout << line << endl << flush;
    }

  delete conn;

  return 0;
}
